### ISSUE TITLE
Support for a generic `InputObject`

### ISSUE THREAD

**Issue by leebenson** [async-graphql/async-graphql#387]
I'm attempting to create a 'base' generic `InputObject`, that could be extended into a concrete type without reimplementing the fields.

Something like this would be ideal:

```rust
#[derive(Default)]
pub struct EqualityFilter<T: PartialEq + Eq> {
    equals: Option<T>,
    not_equals: Option<T>,
}

#[InputObject(name = "SomeConcreteName")]
impl EqualityFilter<SomeType> {}
```

Is there an existing pattern that might support something close to this approach?

I could write a macro for this specific use-case, but curious if there's an existing way to do this that I might be overlooking.

Thanks!

**Comment by sunli829** [comment]
There is currently no method that supports the generic `InputObject`.

Because the following code cannot get the fields of the struct, it cannot be done.

```rust
#[InputObject(name = "SomeConcreteName")]
impl EqualityFilter<SomeType> {}
```

I am trying the following implementation.

```rust
#[derive(Default)]
#[graphql(generic(name = "SomeConcreteName", params("SomeType")))]
pub struct EqualityFilter<T: PartialEq + Eq> {
    equals: Option<T>,
    not_equals: Option<T>,
}
```

**Comment by sunli829** [comment]
Released in `v2.4.8` 😁

Here is the test: 
https://github.com/async-graphql/async-graphql/blob/4a4e4e7219d77ef379a535685cb83d32c45f85be/tests/input_object.rs#L354

When you are free, can you help me add a tutorial about generic `InputObject` in the Book?

_Note:  `params` is a list._
```rust
 #[derive(InputObject)]
#[graphql(concrete(name = "SomeConcreteName", params(A, B, C)))]
```

**Comment by leebenson** [comment]
Thanks, as usual, for being so attentive to feature requests! Really appreciated.

I'd be happy to write up a doc page for this.

There's one thing I'm not clear on -- when adding multiple type annotations, the schema that's generated seems off.

I'm attempting to use this:

```rust
#[derive(Enum, Copy, Clone, Eq, PartialEq)]
/// Type of component
pub enum ComponentType {
    Source,
    Transform,
    Sink,
}

#[derive(InputObject)]
#[graphql(concrete(name = "ComponentTypeEqualityFilter", params(ComponentType)))]
#[graphql(concrete(name = "StringEqualityFilter", params(String)))]
pub struct EqualityFilter<T: InputType + PartialEq + Eq> {
    equals: Option<T>,
    not_equals: Option<T>,
}

impl<T: InputType + PartialEq + Eq> EqualityFilter<T> {
    pub fn filter_value(&self, value: T) -> bool {
        filter_check!(
            // Equals
            self.equals.as_ref().map(|s| value.eq(s)),
            // Not equals
            self.not_equals.as_ref().map(|s| !value.eq(s))
        );
        true
    }
}
```

And then implementing both as fields of another `InputObject`, like so:

```rust
#[derive(Default, InputObject)]
pub struct ComponentsFilter {
    name: Option<Vec<filter::StringFilter>>,
    component_type: Option<Vec<filter::EqualityFilter<ComponentType>>>,
    big_test: Option<Vec<filter::EqualityFilter<String>>>,
    and: Option<Vec<Self>>,
    or: Option<Vec<Self>>,
}
```

But looking at the generated schema, this is what I get:

```gql
input ComponentsFilter {
  name: [StringFilter!]
  componentType: [EqualityFilter!] # <-- shouldn't this be [ComponentTypeEqualityFilter!]?
  bigTest: [EqualityFilter!] # <-- and this one [StringEqualityFilter!]?
  and: [ComponentsFilter!]
  or: [ComponentsFilter!]
}
```

Both fields appear to be `[EqualityFilter!]`, which is implemented as `String` -- and not `ComponentType`:

```gql
input EqualityFilter {
  equals: String
  notEquals: String
}
```

I might be implementing it incorrectly -- would you mind taking a quick look? Thanks!

**Comment by sunli829** [comment]
> Thanks, as usual, for being so attentive to feature requests! Really appreciated.
> 
> I'd be happy to write up a doc page for this.
> 
> There's one thing I'm not clear on -- when adding multiple type annotations, the schema that's generated seems off.
> 
> I'm attempting to use this:
> 
> ```rust
> #[derive(Enum, Copy, Clone, Eq, PartialEq)]
> /// Type of component
> pub enum ComponentType {
>     Source,
>     Transform,
>     Sink,
> }
> 
> #[derive(InputObject)]
> #[graphql(concrete(name = "ComponentTypeEqualityFilter", params(ComponentType)))]
> #[graphql(concrete(name = "StringEqualityFilter", params(String)))]
> pub struct EqualityFilter<T: InputType + PartialEq + Eq> {
>     equals: Option<T>,
>     not_equals: Option<T>,
> }
> 
> impl<T: InputType + PartialEq + Eq> EqualityFilter<T> {
>     pub fn filter_value(&self, value: T) -> bool {
>         filter_check!(
>             // Equals
>             self.equals.as_ref().map(|s| value.eq(s)),
>             // Not equals
>             self.not_equals.as_ref().map(|s| !value.eq(s))
>         );
>         true
>     }
> }
> ```
> 
> And then implementing both as fields of another `InputObject`, like so:
> 
> ```rust
> #[derive(Default, InputObject)]
> pub struct ComponentsFilter {
>     name: Option<Vec<filter::StringFilter>>,
>     component_type: Option<Vec<filter::EqualityFilter<ComponentType>>>,
>     big_test: Option<Vec<filter::EqualityFilter<String>>>,
>     and: Option<Vec<Self>>,
>     or: Option<Vec<Self>>,
> }
> ```
> 
> But looking at the generated schema, this is what I get:
> 
> ```graphql
> input ComponentsFilter {
>   name: [StringFilter!]
>   componentType: [EqualityFilter!] # <-- shouldn't this be [ComponentTypeEqualityFilter!]?
>   bigTest: [EqualityFilter!] # <-- and this one [StringEqualityFilter!]?
>   and: [ComponentsFilter!]
>   or: [ComponentsFilter!]
> }
> ```
> 
> Both fields appear to be `[EqualityFilter!]`, which is implemented as `String` -- and not `ComponentType`:
> 
> ```graphql
> input EqualityFilter {
>   equals: String
>   notEquals: String
> }
> ```
> 
> I might be implementing it incorrectly -- would you mind taking a quick look? Thanks!

Please wait, I am confirming whether this problem really exists.

**Comment by leebenson** [comment]
Ah, hold on, I see you changed the comment to 2.4.9 and not 2.4.8-- I only tested with 2.4.8. Trying now... will get back to you in a few mins!

**Comment by leebenson** [comment]
Hmm yeah, this still appears to be happening on 2.4.9.

**Comment by sunli829** [comment]
I confirmed that this problem does exist. . . .

**Comment by sunli829** [comment]
I fixed this in `v2.4.10`. 🙂 @leebenson

**Comment by leebenson** [comment]
Beautiful, thanks so much!
